src/fv/share/tdyugrid.c

```diff
@@ -0,0 +1,1209 @@
+      PetscInt vertID = (PetscInt) v_ptr[icell*stride + vertex_offset + ivertex];
+      if (vertID > 0) {
+        numVertices++;
+      }
+    }
+  }
+
+  // Now save the vertex IDs
+  PetscInt IntArray1[numVertices];
+  numVertices = 0;
+  for (PetscInt icell=0; icell<ngmax; icell++) {
+    for (PetscInt ivertex=0; ivertex<max_nvert; ivertex++) {
+      PetscInt vertID = (PetscInt) v_ptr[icell*stride + vertex_offset + ivertex];
+      if (vertID > 0) {
+        IntArray1[numVertices++] = vertID;
+        v_ptr[icell*stride + vertex_offset + ivertex] = numVertices;
+      }
+    }
+  }
+  ierr = VecRestoreArray(*LocalOrderVec, &v_ptr); CHKERRQ(ierr);
+
+  PetscInt IntArray2[numVertices];
+  PetscInt IntArray3[numVertices];
+  PetscInt IntArray4[numVertices];
+  for (PetscInt ivertex=0; ivertex<numVertices; ivertex++) {
+    IntArray1[ivertex] = IntArray1[ivertex] - 1;
+    IntArray2[ivertex] = ivertex;
+    IntArray3[ivertex] = -1;
+    IntArray4[ivertex] = -1;
+  }
+
+  ierr = PetscSortIntWithPermutation(numVertices, IntArray1, IntArray2); CHKERRQ(ierr);
+
+  PetscInt count=0;
+  PetscInt idx=IntArray2[count];
+  IntArray3[0] = IntArray1[idx];
+  IntArray4[idx] = count;
+
+  for (PetscInt ivertex=0; ivertex<numVertices; ivertex++) {
+    PetscInt idx=IntArray2[ivertex];
+    PetscInt vertID = IntArray1[idx];
+    if (vertID > IntArray3[count]) {
+      count++;
+      IntArray3[count] = vertID;
+    }
+    IntArray4[IntArray2[ivertex]] = count;
+  }
+
+  numVertices = count+1;
+  *NewNumVertices = numVertices;
+
+  // Save vertex ids in natural-order
+  ierr = TDyAllocate_IntegerArray_1D(&ugrid->vertex_ids_natural, numVertices); CHKERRQ(ierr);
+  for (PetscInt ivertex=0; ivertex<numVertices; ivertex++) {
+    ugrid->vertex_ids_natural[ivertex] = IntArray3[ivertex];
+  }
+
+  ierr = TDyAllocate_IntegerArray_1D(&ugrid->cell_num_vertices, ngmax); CHKERRQ(ierr);
+  ierr = VecGetArray(*LocalOrderVec, &v_ptr); CHKERRQ(ierr);
+
+  // The previously allocated memoroy for ugrid->cell_vertices cannot be deallocated
+  // because the dimension for the memory was prior to mesh partition and after
+  // mesh partition the size of ugrid->cell_vertices is not known
+  ierr = TDyAllocate_IntegerArray_2D(&ugrid->cell_vertices, ngmax, max_nvert); CHKERRQ(ierr);
+
+  for (PetscInt icell=0; icell<ngmax; icell++) {
+
+    PetscInt count=0;
+    ugrid->cell_num_vertices[icell] = count;
+
+    for (PetscInt ivertex=0; ivertex<max_nvert; ivertex++) {
+      PetscInt vertID = (PetscInt) v_ptr[icell*stride + vertex_offset + ivertex];
+      if (vertID > 0) {
+        ugrid->cell_vertices[icell][count++] = IntArray4[vertID-1];
+        ugrid->cell_num_vertices[icell] = count;
+        v_ptr[icell*stride + vertex_offset + ivertex] = IntArray4[vertID-1]+1;
+      }
+    }
+  }
+
+  ierr = VecRestoreArray(*LocalOrderVec, &v_ptr); CHKERRQ(ierr); // Now vertex ids are in local-order
+
+#ifdef UGRID_DEBUG
+  char filename[30];
```

Comment:
I recommend using something bigger than 30 here, and using `snprintf` instead of `sprintf` to make sure you don't overflow your string.

Reply:
I used `char filename[PETSC_MAX_PATH_LEN]`